File: frontend/src/types/learning.ts
// ============================================================
// 学习成长类型定义（二期 M8）
// ============================================================

export interface StudyPlan {
  id: string
  name: string
  description: string | null
  target_date: string | null
  estimated_hours: number | null
  progress: number
  status: 'active' | 'paused' | 'completed'
  config: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface Flashcard {
  id: string
  plan_id: string | null
  front: string
  back: string
  card_type: 'qa' | 'cloze' | 'image'
  category: string | null
  tags: string[]
  difficulty: number  // 1-5
  ef: number  // SM-2 难度因子，下限1.3
  interval: number  // 天
  repetition: number
  next_review: string | null
  last_reviewed_at: string | null
  review_count: number
  created_at: string
  updated_at: string
}

export interface StudyTimeLog {
  id: string
  plan_id: string | null
  subject: string | null
  duration: number  // 分钟
  note: string | null
  source: 'manual' | 'timer' | 'auto'
  logged_date: string
  created_at: string
}

export interface ReviewResult {
  card_id: string
  quality: number  // 0-5
  next_interval: number
  next_review: string
}

// ============================================================
// S6-1 领域规则引擎：从目标生成学习计划
// 全程离线可用，不依赖 API Key
// ============================================================

/** 计划中的一个能力单元（含阶段：1 打基础 / 2 核心练习 / 3 整合实战） */
export interface GeneratedUnit {
  name: string
  phase: number
  task_count: number
}

/** 计划中的一条任务。day 在周模式下表示周序号 */
export interface GeneratedPlanItem {
  unit_id: string
  day: number
  title: string
  task: string
  duration: string
  /** 由 duration 文本归一得到的分钟数（后端已解析，无法识别时为 null） */
  estimated_minutes: number | null
  /** 达标标准：做到什么程度算完成。领域库 90/90 条任务都有 */
  acceptance: string
  is_verify: boolean
}

export interface GeneratedPlan {
  goal: string
  skill_name: string
  /** false 表示未命中领域库，走了「入门准备→基础操作→核心练习→实战产出」通用兜底 */
  matched: boolean
  domain_key: string | null
  domain_name: string | null
  total_days: number
  minutes_per_day: number
  /** true 表示按周排期（周期 > 45 天） */
  weekly_mode: boolean
  verify_task: string
  units: GeneratedUnit[]
  items: GeneratedPlanItem[]
  item_count: number
}

export interface GeneratePlanParams {
  goal: string
  total_days?: number
  minutes_per_day?: number
}

export interface ApplyPlanParams extends GeneratePlanParams {
  plan_name?: string
  project_id?: string
  start_date?: string
}

export interface ApplyPlanResult {
  plan: StudyPlan
  task_count: number
  task_ids: string[]
  preview: GeneratedPlan
}

File: backend/app/schemas/learning.py
# ============================================================
# 学习成长 schema（二期 M8）
# ============================================================
from __future__ import annotations

from datetime import date, datetime
from pydantic import BaseModel, Field


class StudyPlanOut(BaseModel):
    id: str
    name: str
    description: str | None = None
    target_date: date | None = None
    estimated_hours: int | None = None
    progress: int = 0
    status: str = "active"
    config: dict = {}
    created_at: datetime
    updated_at: datetime


class CreateStudyPlanRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    target_date: date | None = None
    estimated_hours: int | None = None


class GeneratePlanRequest(BaseModel):
    """从自然语言目标生成学习计划（S6-1 · **离线可用**）。

    走内置领域知识库（13 领域 / 52 能力单元 / 90 条带达标标准的任务）；
    未命中领域时退化为「入门准备 → 基础操作 → 核心练习 → 实战产出」
    四阶段通用拆解。**全程不依赖任何模型，也不需要 API Key。**
    """

    goal: str = Field(min_length=1, max_length=200, description="学习目标，如「我想学雅思」")
    total_days: int = Field(default=30, ge=1, le=730, description="计划周期（天）")
    minutes_per_day: int = Field(default=45, ge=10, le=600, description="每天可投入分钟数")


class ApplyPlanRequest(GeneratePlanRequest):
    """生成计划并落成任务（写入 Task 表，可参与日程与统计）。"""

    plan_name: str | None = Field(default=None, max_length=200, description="学习计划名称")
    project_id: str | None = Field(default=None, description="将生成的任务归入该项目")
    start_date: date | None = Field(default=None, description="起始日期，默认今天")


class FlashcardOut(BaseModel):
    id: str
    plan_id: str | None = None
    front: str
    back: str
    card_type: str = "qa"
    category: str | None = None
    tags: list = []
    difficulty: int = 3
    ef: float = 2.5
    interval: int = 0
    repetition: int = 0
    next_review: date | None = None
    last_reviewed_at: datetime | None = None
    review_count: int = 0
    created_at: datetime
    updated_at: datetime


class CreateFlashcardRequest(BaseModel):
    front: str = Field(min_length=1)
    back: str = Field(min_length=1)
    plan_id: str | None = None
    card_type: str = "qa"
    category: str | None = None
    tags: list = []
    difficulty: int = 3


class ReviewSubmitRequest(BaseModel):
    quality: int = Field(ge=0, le=5)


class StudyTimeLogOut(BaseModel):
    id: str
    plan_id: str | None = None
    subject: str | None = None
    duration: int
    note: str | None = None
    source: str | None = None
    logged_date: date
    created_at: datetime


class CreateStudyTimeLogRequest(BaseModel):
    duration: int = Field(gt=0)
    subject: str | None = None
    plan_id: str | None = None
    note: str | None = None
    source: str = "manual"
    logged_date: date | None = None
